crates/client/db/src/bonsai_db.rs

```diff
@@ -148,144 +160,121 @@ impl BonsaiDatabase for BonsaiDb<'_> {
-//         let handle = self.db.get_column(self.column_mapping.map(key));
-//         Ok(self.txn.get_cf(&handle, key.as_slice()).map(|value| value.is_some())?)
-//     }
-
-//     fn insert(
-//         &mut self,
-//         key: &DatabaseKey,
-//         value: &[u8],
-//         batch: Option<&mut Self::Batch>,
-//     ) -> Result<Option<Vec<u8>>, Self::DatabaseError> {
-//         tracing::trace!("Inserting into RocksDB: {:?} {:?}", key, value);
-//         let handle = self.db.get_column(self.column_mapping.map(key));
-//         let old_value = self.txn.get_cf(&handle, key.as_slice())?;
-//         if let Some(batch) = batch {
-//             batch.put_cf(&handle, key.as_slice(), value);
-//         } else {
-//             self.txn.put_cf(&handle, key.as_slice(), value)?;
-//         }
-//         Ok(old_value)
-//     }
-
-//     fn remove(
-//         &mut self,
-//         key: &DatabaseKey,
-//         batch: Option<&mut Self::Batch>,
-//     ) -> Result<Option<Vec<u8>>, Self::DatabaseError> {
-//         tracing::trace!("Removing from RocksDB: {:?}", key);
-//         let handle = self.db.get_column(self.column_mapping.map(key));
-//         let old_value = self.txn.get_cf(&handle, key.as_slice())?;
-//         if let Some(batch) = batch {
-//             batch.delete_cf(&handle, key.as_slice());
-//         } else {
-//             self.txn.delete_cf(&handle, key.as_slice())?;
-//         }
-//         Ok(old_value)
-//     }
-
-//     fn remove_by_prefix(&mut self, prefix: &DatabaseKey) -> Result<(), Self::DatabaseError> {
-//         tracing::trace!("Getting from RocksDB: {:?}", prefix);
-//         let handle = self.db.get_column(self.column_mapping.map(prefix));
-//         let iter = self.txn.iterator_cf(&handle, IteratorMode::From(prefix.as_slice(), Direction::Forward));
-//         let mut batch = self.create_batch();
-//         for kv in iter {
-//             if let Ok((key, _)) = kv {
-//                 if key.starts_with(prefix.as_slice()) {
-//                     batch.delete_cf(&handle, &key);
-//                 } else {
-//                     break;
-//                 }
-//             } else {
-//                 break;
-//             }
-//         }
-//         drop(handle);
-//         self.write_batch(batch)?;
-//         Ok(())
-//     }
-
-//     fn write_batch(&mut self, batch: Self::Batch) -> Result<(), Self::DatabaseError> {
-//         Ok(self.txn.rebuild_from_writebatch(&batch)?)
-//     }
-// }
-
-impl<'db> BonsaiPersistentDatabase<BasicId> for BonsaiDb<'db>
-where
-    Self: 'db,
-{
-    type Transaction = Self;
+fn to_changed_key(k: &DatabaseKey) -> (usize, ByteVec) {
+    (
+        match k {
+            DatabaseKey::Trie(_) => 0,
+            DatabaseKey::Flat(_) => 1,
+            DatabaseKey::TrieLog(_) => 2,
+        },
+        k.as_slice().into(),
+    )
+}
+
+pub struct BonsaiTransaction {
+    snapshot: Arc<SnapshotWithDBArc<DB>>,
```

Comment:
that's because the snapshot needs to live in the Snapshots struct which holds and manage them, but we don't want to hold a lock the whole time a snapshot is used in a bonsai transaction
so, instead, the Snapshots struct will hold the lock only for the time of cloning the Arc to the snapshot

in addition, the snapshot holding an Arc of the database is very important for safety reasons - this ensures that the snapshot cannot outlive the db. Otherwise the SnapshotWithDBArc should be marked unsafe.

these two arcs have very different goals.

Reply:
This comment in the code would be very useful.

Reply:
done :)